scripts/coverage_diff.py

```diff
@@ -0,0 +1,96 @@
+# https://fpgasoftware.intel.com/
+
+# Example usage: python scripts/coverage_diff.py child_build_dir parent_build_dir
+import xml.etree.ElementTree as ET
+import sys
+import argparse
+import glob
+
+parser = argparse.ArgumentParser(description='Optional app description')
+parser.add_argument('child_coverage', type=str,
+                    help='Child branch build directory')
+parser.add_argument('parent_coverage', type=str,
+                    help='Parent branch build directory')
+args = parser.parse_args()
+
+child_cov_file = glob.glob(args.child_coverage + "/Testing/**/Coverage.xml", recursive = True)[0]
+parent_cov_file = glob.glob(args.parent_coverage + "/Testing/**/Coverage.xml", recursive = True)[0]
+
+def want_file(fn):
+    return fn.startswith("./src")
+
+def pprint_table(matrix):
+    s = [[str(e) for e in row] for row in matrix]
+    lens = [max(map(len, col))+1 for col in zip(*s)]
+    fmt = '\t'.join('{{:{}}}'.format(x) for x in lens)
+    table = [fmt.format(*row) for row in s]
+    print('\n'.join(table))
+
+with open(child_cov_file, 'r') as f:
+    child_data = f.read()
+with open(parent_cov_file, 'r') as f:
+    parent_data = f.read()
+
+# Parse child
+child_file_to_coverage = {}
+root = ET.fromstring(child_data)
+for elm in root.findall(".//File[@FullPath]"):
+    fn = elm.attrib["FullPath"]
+    if not want_file(fn):
+        continue
+    percent_cov = float(elm.find('PercentCoverage').text)
+    child_file_to_coverage[fn] = percent_cov
+
+# Parse parent
+parent_file_to_coverage = {}
+root = ET.fromstring(parent_data)
+for elm in root.findall(".//File[@FullPath]"):
+    fn = elm.attrib["FullPath"]
+    if not want_file(fn):
+        continue
+    percent_cov = float(elm.find('PercentCoverage').text)
+    parent_file_to_coverage[fn] = percent_cov
+
+columns = ["File Name", "Coverage (%)", "Delta Coverage"]
+separator = ['-'*len(cn) for cn in columns]
+result = []
+fail = False
+coverage_decreased = False
+severe_decrease = False
+new_file_not_covered = False
+for fn in child_file_to_coverage:
+    parent_cov = parent_file_to_coverage[fn] if fn in parent_file_to_coverage else 0
+    child_cov = child_file_to_coverage[fn]
+    delta_cov = child_cov - parent_cov
+    result.append([fn, child_cov, delta_cov])
+    if fn not in parent_file_to_coverage and delta_cov <= 20:
+        # New files need to have at least 20% coverage
+        new_file_not_covered = True
+    if delta_cov < 0 and abs(delta_cov) >= min(20, parent_cov):
+        severe_decrease = True
+    if delta_cov < 0:
+        coverage_decreased = True
+
+result = [columns, separator] + result
+
+if new_file_not_covered:
+    pprint_table(result)
+    print("New files are not covered")
+    exit(1)
+elif severe_decrease:
+    pprint_table(result)
+    print("Coverage decreased severely or a file lost coverage.")
+    exit(2)
+elif coverage_decreased:
+    pprint_table(result)
+    print("Coverage decreased.")
+    exit(78)
```

Comment:
This special value would need an explanation, since one would naturally expect `3`.

But see below, exiting with `1` for any kind of failed coverage should be sufficient.

Reply:
The last few step of the workflow behave differently depending on the exit code of this python script, 78 exit code is used to be a "ignore status" for github workflow, although they now have removed it. 

The purpose of separating exit code is to let the workflow skip certain step. The workflow succeed but the step will look different than passing, just as a warning that the coverage has slightly decreased.

Reply:
I see, thanks for the explanation of exit status 78. If that is no longer used, the only purpose of the different exit codes handling is to display different messages, correct? I still think a single workflow step, simply the running the script, would sufficient and more maintenable. Currently the script and workflow handle each case *three* times, where one time per case seems sufficient. Other than a pass or failed, there is no other result for a workflow, correct?

Reply:
Right that make sense. I originally did it only for the "skip" symbol beside slight coverage decrease case. But I guess that doesn't bring too much value in terms of warning the developers. I have removed the extra workflow steps at the end.

Reply:
> The workflow succeed but the step will look different than passing, just as a warning that the coverage has slightly decreased.

I missed this bit about the warning, now I understand your original intention. Unfortunately I would think that regardless of whether the warning is printed by the script, or shown as a separate workflow step, it would still be ignored. As long as the workflow shows a green check mark, I don't think anyone would check each time, unless they were specifically working on the workflow itself.

What do you think, would it be straight-forward to post a comment to the pull request from the python script with the coverage diff? That way the submitter has to acknowledge warnings, and also receives positive feedback when the coverage increases.

Reply:
That's a great idea! Since PRs are a flavor of issue, the [issue comment API](https://docs.github.com/en/rest/reference/issues#create-an-issue-comment) will work

Reply:
Great :slightly_smiling_face: For GitHub API usage, take a look at the different options and let me know which one(s) you prefer. A straight-forward option would be one of the python modules shipped with Ubuntu 20.04, e.g., [python3-github](https://packages.ubuntu.com/focal/python3-github) for high-level or [python3-requests](https://packages.ubuntu.com/focal/python3-requests) for low-level usage, but there might be better options.